### PR TITLE
ORTModule - FastGeluFusion/fp16 fix and minor LayerNormFusion cleanup

### PR DESCRIPTION
**Pull request by ytaous**
**Description**: During the GPT2 model benchmark using ORTModule, found FastGeluFusion doesn't happen when fp16 is enabled. The reason is due to extra Cast nodes exported from PyTorch, which break the matching patterns.

The fix is to add extra pattern logic in the second formula to conver this case.
With the fix, the throughput for GPT2 is improved another ~8% comparing to the test without the fix.

108146746-956dea00-7082-11eb-95ab-9ba4126ed44f

### CODE REVIEW

File: onnxruntime/core/optimizer/fast_gelu_fusion.cc [hunk 1]
```diff
@@ -141,6 +141,24 @@ MatchResult FastGeluFusion::CheckSecondFormula(Graph& graph, Node& pow1_node,
   }
   nodes_to_fuse.push_back(add1_node);
 
+  // check if pow node has Cast parent, expect Add has same Cast parent as well
+  const Node* p_cast1_node = graph_utils::FirstParentByType(pow1_node, "Cast");
+  if (p_cast1_node != nullptr) {
+    Node& cast1_node = *graph.GetNode(p_cast1_node->Index());
+    // this is fused Cast node, so expect more than 1 output edges
+    if (!CheckNode(graph, cast1_node, "Cast", {9, 13}, pow1_node.GetExecutionProviderType(), false)){
```
Comment (ashbhandare) [inline onnxruntime/core/optimizer/fast_gelu_fusion.cc:149]: can you assert that the output edges count is 2

**Comment by ashbhandare** [comment]
I think we are missing the check to see if the cast is to a higher precision. Can we check if the 'to' attribute of the cast is to FP32?

File: onnxruntime/core/optimizer/fast_gelu_fusion.cc [hunk 2]
```diff
@@ -201,6 +238,30 @@ Status FastGeluFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
...
+      if (p_cast1_node != nullptr) {
+        // we've done the node check in second formula for pow node
+        Node& cast1_node = *graph.GetNode(p_cast1_node->Index());
+        cast_input_arg = cast1_node.MutableInputDefs()[0];
+
+        const Node* p_cast3_node = graph_utils::FirstParentByType(mul5_node, "Cast");
+        if (p_cast3_node == nullptr) continue;
+
+        Node& cast3_node = *graph.GetNode(p_cast3_node->Index());
+        if (!CheckNode(graph, cast3_node, "Cast", {9, 13}, node.GetExecutionProviderType(), true)) {
```
Comment (ashbhandare) [inline onnxruntime/core/optimizer/fast_gelu_fusion.cc:255]: output count 1 check

Comment (ytaous) [reply onnxruntime/core/optimizer/fast_gelu_fusion.cc:255]: output 1 check is already covered by CheckNode() with true param

---
In reply to: [578681548](https://github.com/microsoft/onnxruntime/pull/6734#discussion_r578681548) [](ancestors = 578681548)

**Comment by ytaous** [comment]
Per discussion - the main concern is between Mul and Cast, but since Mul has to be 0.5, which is < 1, so no concern of NaNs.
thus we can skip the attribute check.

---
In reply to: [781573835](https://github.com/microsoft/onnxruntime/pull/6734#issuecomment-781573835) [](ancestors = 781573835)

**Review by ashbhandare — APPROVED** [review]

**Pull request merged**